sklearn/metrics/cluster/supervised.py

```diff
@@ -870,3 +873,68 @@ def entropy(labels):
     # log(a / b) should be calculated as log(a) - log(b) for
     # possible loss of precision
     return -np.sum((pi / pi_sum) * (np.log(pi) - log(pi_sum)))
+
+
+def class_cluster_match(y_true, y_pred):
+    """Translate prediction labels to maximize the accuracy.
+
+    Translate the prediction labels of a clustering output to enable calc
+    of external metrics (eg. accuracy, f1_score, ...). Translation is done by
+    maximization of the confusion matrix :math:`C` main diagonal sum
+    :math:`\sum{i=0}^{K}C_{i, i}`. Notice the number of cluster has to be equal
+     or smaller than the number of true classes.
+
+    Parameters
+    ----------
+    y_true : array, shape = [n_samples]
+        Ground truth (correct) target values.
+    y_pred : array, shape = [n_samples]
+        Estimated targets as returned by a clustering algorithm.
+
+    Returns
+    -------
+    trans : array, shape = [n_classes, n_classes]
+        Mapping of y_pred clusters, such that :math:`trans\subseteq y_true`
+
+    References
+    ----------
+
+    Examples
+    --------
+    >>> from sklearn.metrics import confusion_matrix
+    >>> from sklearn.metrics.cluster import class_cluster_match
+    >>> y_true = ["class1", "class2", "class3", "class1", "class1", "class3"]
+    >>> y_pred = [0, 0, 2, 2, 0, 2]
+    >>> y_pred_translated = class_cluster_match(y_true, y_pred)
+    >>> y_pred_translated
+    ['class1', 'class1', 'class3', 'class3', 'class1', 'class3']
+    >>> confusion_matrix(y_true, y_pred_translated)
+    array([[2, 0, 1],
+           [1, 0, 0],
+           [0, 0, 2]])
+    """
+
+    classes = unique_labels(y_true).tolist()
+    n_classes = len(classes)
+    clusters = unique_labels(y_pred).tolist()
+    n_clusters = len(clusters)
+
+    if n_clusters > n_classes:
+        classes += ['DEF_CLASS'+str(i) for i in range(n_clusters-n_classes)]
```

Comment:
I don't think this should be necessary. The contingency matrix will come out non-square, and then we only need to add rows or columns of zeros. In terms of the output, see my "default_label" comment above.